Improve Version Parsing to handle more separators

edgars-cirulis opened the pull request:
This pull request updates the parse method in the Version class to improve its handling of version strings. The previous implementation used the '.' separator to split the version string into sections, but this approach is not robust as there may be other separators present in version strings. The new implementation uses a lambda function to check for changes in character class (digit or non-digit) or the presence of '-' or '+' characters as separators. This approach is more robust and can handle a wider range of version string formats. This change will make the Version class more reliable and easier to use.

Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

Vaxry committed:
Version.cpp: Improve version parsing to handle mixed numeric and alphabetic characters

Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

Ryex commented:
as for addressing the FIXME...
IMO it should also split every time it changes over between alpha and numeric categories and if it sees `-` or `_`

so 22w43a would count as `[ "22", "w", "43", "a"]`

that would make snapshot versions more order-able
it's not perfect of course but I think it's serviceable.

but
with how you have this written I'm fairly sure `1.5.4` gets split into `["1", ".5", "5", ".4", "4"]`
because of the nested capture groups.
not sure that is ideal or intended.

I'm fairly sure regex is just a bad tool for this

I think it would be better to write a simple parser that loops though the string character by character. Each time it encounters one of `[\.\-_]` or the capture class (`\d` or `[a-zA-Z]`) switches it appends the last complete section and starts on a new section.

edgars-cirulis commented:
> as for addressing the FIXME... IMO it should also split every time it changes over between alpha and numeric categories and if it sees `-` or `_`
> 
> so 22w43a would count as `[ "22", "w", "43", "a"]`
> 
> that would make snapshot versions more order-able it's not perfect of course but I think it's serviceable.
> 
> but with how you have this written I'm fairly sure `1.5.4` gets split into `["1", ".5", "5", ".4", "4"]` because of the nested capture groups. not sure that is ideal or intended.
> 
> I'm fairly sure regex is just a bad tool for this
> 
> I think it would be better to write a simple parser that loops though the string character by character. Each time it encounters one of `[\.\-_]` or the capture class (`\d` or `[a-zA-Z]`) switches it appends the last complete section and starts on a new section.

You are correct that the regex approach may not work as intended, especially for version strings that contain both numeric and alphabetic characters. A simple parser that loops through the string character by character, as you suggested, would be a better approach to handle this use case.

This way, the version string "22w43a" would be split into ["22", "w", "43", "a"] as you expected.

edgars-cirulis commented:
Sorry for the force push. The previous commit was https://github.com/PrismLauncher/PrismLauncher/commit/801188681613639f0ce95cf3237ce401a54ac237

Ryex commented on launcher/Version.cpp:
```diff
@@ -74,12 +74,36 @@ bool Version::operator!=(const Version &other) const
+    bool lastCharWasDigit = false;
+    for (int i = 0; i < m_string.size(); ++i) {
+        if(m_string[i].isDigit()){
+            if(!lastCharWasDigit){
```
Hmm I think the logic could be cleaned up significantly but we're getting there...

```c++

auto class_change = [] (QChar lastChar, QChar currentChar) {
    return (( lastChar.isLetter() && currentChar.isDigit() ) || (lastChar.isDigit() && currentChar.isLetter()) );
} 
```

Using an anonymous function like this would prevent the code duplication and imo make the logic clearer.

Edgars-Cirulis replied:
done in https://github.com/PrismLauncher/PrismLauncher/pull/751/commits/f2ad6c5eef3e54206257dfd92369b8e49954fd79

edgars-cirulis replied:
done in https://github.com/PrismLauncher/PrismLauncher/pull/751/commits/b58a9b32f65558ff61ed91259b309c2ece2ddab8

Ryex commented on launcher/Version.cpp:
```diff
@@ -74,12 +74,36 @@ bool Version::operator!=(const Version &other) const
+            currentSection += m_string[i];
+            lastCharWasDigit = false;
+        }
+        else if(m_string[i] == '-' || m_string[i] == '_'){
```
Also, you forgot to still split on '.'

edgars-cirulis replied:
done in https://github.com/PrismLauncher/PrismLauncher/pull/751/commits/fa0c3790a618853a6708128f4cbf04cbe6910a4f

Ryex reviewed (changes requested):
Looking good but it could be clearer. 👍

Vaxry committed:
Version.cpp: Add version string parser to split on '.' character

Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

Vaxry committed:
Version.cpp: Use anonymous function to eliminate code duplication

Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

flowln commented on launcher/Version.cpp:
```diff
@@ -74,12 +74,28 @@ bool Version::operator!=(const Version &other) const
+        return (( lastChar.isLetter() && currentChar.isDigit() ) || (lastChar.isDigit() && currentChar.isLetter()) );
+    };
+    for (int i = 0; i < m_string.size(); ++i) {
+        if(m_string[i].isDigit() || m_string[i].isLetter()){
```
you can substitute all the `m_string[i]` for a line at the beginning of the loop with `const auto& current_char = m_string.at(i);`

edgars-cirulis replied:
done in https://github.com/PrismLauncher/PrismLauncher/pull/751/commits/5d558702706e9ad464340879ec8a67c73f1892a9

flowln commented on launcher/Version.cpp:
```diff
@@ -74,12 +74,28 @@ bool Version::operator!=(const Version &other) const
+        return (( lastChar.isLetter() && currentChar.isDigit() ) || (lastChar.isDigit() && currentChar.isLetter()) );
+    };
+    for (int i = 0; i < m_string.size(); ++i) {
+        if(m_string[i].isDigit() || m_string[i].isLetter()){
```
please run `clang-format` on here

edgars-cirulis replied:
done in https://github.com/PrismLauncher/PrismLauncher/pull/751/commits/624a59631eea9ab834f430f2ca259cbe8c6d3c57

flowln reviewed (commented):
considering you're already messing with the version decomposition here, you could might as well try following [`FlexVer decomposition`](https://github.com/unascribed/FlexVer/blob/trunk/SPEC.md#decomposition) to try avoiding fragmentation in the version comparison format between applications (not that it will solve all the problems, but it's a step in the right direction IMO). It would also give us a few cool test instances we could test the implementation in :-)

Vaxry committed:
Version.cpp: Simplify Version::parse by using const auto& current_char

Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

Vaxry committed:
Version.cpp: Format parse function code using clang-format

Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

edgars-cirulis commented:
> considering you're already messing with the version decomposition here, you could might as well try following [`FlexVer decomposition`](https://github.com/unascribed/FlexVer/blob/trunk/SPEC.md#decomposition) to try avoiding fragmentation in the version comparison format between applications (not that it will solve all the problems, but it's a step in the right direction IMO). It would also give us a few cool test instances we could test the implementation in :-)

 I have  tried FlexVer decomposition and found it to be too messy. Therefore, I have decided to stick with my current implementation as it works. However, in the future, if I decide to revisit this issue, FlexVer decomposition could be a step in the right direction for avoiding fragmentation in the version comparison format between applications.

Ryex commented:
> > considering you're already messing with the version decomposition here, you could might as well try following [`FlexVer decomposition`](https://github.com/unascribed/FlexVer/blob/trunk/SPEC.md#decomposition) to try avoiding fragmentation in the version comparison format between applications (not that it will solve all the problems, but it's a step in the right direction IMO). It would also give us a few cool test instances we could test the implementation in :-)
> 
> I have tried FlexVer decomposition and found it to be too messy. Therefore, I have decided to stick with my current implementation as it works. However, in the future, if I decide to revisit this issue, FlexVer decomposition could be a step in the right direction for avoiding fragmentation in the version comparison format between applications.

could you elaborate on what you mean by too messy?

as I understand flexver it should be as simple as doing this

```c++
    auto classChange = [](QChar lastChar, QChar currentChar) {
        return !lastChar.isNull() && ((!lastChar.isDigit() && currentChar.isDigit()) || (lastChar.isDigit() && !currentChar.isDigit()));
    };
```

then checking the first character in a section for specials like `.`, `-`, `_`, and `+`

edgars-cirulis commented:
> 

Ohh. Well 

> > > considering you're already messing with the version decomposition here, you could might as well try following [`FlexVer decomposition`](https://github.com/unascribed/FlexVer/blob/trunk/SPEC.md#decomposition) to try avoiding fragmentation in the version comparison format between applications (not that it will solve all the problems, but it's a step in the right direction IMO). It would also give us a few cool test instances we could test the implementation in :-)
> > 
> > 
> > I have tried FlexVer decomposition and found it to be too messy. Therefore, I have decided to stick with my current implementation as it works. However, in the future, if I decide to revisit this issue, FlexVer decomposition could be a step in the right direction for avoiding fragmentation in the version comparison format between applications.
> 
> could you elaborate on what you mean by too messy?
> 
> as I understand flexver it should be as simple as doing this
> 
> ```c++
>     auto classChange = [](QChar lastChar, QChar currentChar) {
>         return !lastChar.isNull() && ((!lastChar.isDigit() && currentChar.isDigit()) || (lastChar.isDigit() && !currentChar.isDigit()));
>     };
> ```
> 
> then checking the first character in a section for specials like `.`, `-`, `_`, and `+`

I attempted to implement the FlexVer structure for decomposition, but found that it required significant adjustments to other components such as flameapi, among others. which I could be done later.

Vaxry committed:
Version.cpp: Decompose version strings according to flexver

Co-authored-by: Rachel Powers <508861+Ryex@users.noreply.github.com>
Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

Vaxry committed:
Version.cpp: Remove unnecessary QStringList include

Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

Edgars-Cirulis commented:
@flowln @Scrumplex from my side, it seems to be working fine. However, I need some tests from your side to ensure everything is working properly.

edgars-cirulis commented:
@Ryex Unfortunately, your implementation https://github.com/PrismLauncher/PrismLauncher/pull/751/commits/3bec4a80b3de58d31992eda8497a3d099190b92d did not work as I had already anticipated. Removing the commit 'your implementation' passes all tests with all possible Minecraft versions.
Additionally, by removing those who failed, it passed with 0.04 seconds, whereas my implementation passed with 0.02 seconds.

Ryex committed:
fix: proper null padded version comparison

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
@Edgars-Cirulis, the failure cases are from comparisons between incomplete versions. this is because we are not following the flexver comparison function correctly. In flexver the null padding components are always compared as less than.

I think we want the additional rule that null and 0 are equal so that "1.2" and "1.2.0" are equal

https://github.com/Ryex/PrismLauncher/commits/flexver_version has the two commits needed to fix this.
feel free to merge or cherry pick from it.

EDIT: I may have forgotten to include the QDebug header because it mysteriously works without it under Qt6

Ryex committed:
feat: add debug printing for Version

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented on launcher/Version.h:
```diff
@@ -100,9 +106,17 @@ class Version
+            } else if (numValid && other.m_isNull) {
+                return m_numPart != 0;
+            } else if (m_isNull || other.m_isNull) {
+                return false;
```
oh shoot, this is my bad. 
This effectively declares all string version sections equal to a null pad (because this is implemented with inverse logic though `!=`). this should only be true for "." 

```suggestion
                if ((m_stringPart == ".") || (other.m_stringPart == ".")) return false;
                return true;
```

edgars-cirulis replied:
Done https://github.com/PrismLauncher/PrismLauncher/pull/751/commits/f49ad2ee03974c9fe94882d99d1a2bee67b87285

Vaxry committed:
Version.h: Fix comparison of null version in Version class

Co-authored-by: Rachel Powers <508861+Ryex@users.noreply.github.com>
Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

Vaxry committed:
Version.cpp: Add new line

Signed-off-by: Edgars Cīrulis <edgarsscirulis@gmail.com>

edgars-cirulis commented:
I'm assuming it's done my own tests passed successfully. Now need review.

flowln committed:
feat(tests): add FlexVer test vector to the Version tests

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
refactor: simplify Version operators

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
refactor: clean up Section struct

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat+fix(Version): make comparsion FlexVer-compatible

... and fixes a minor issue in the parsing.

This changes the expected behavior of Versions in one significant way:
Now, Versions like 1.2 or 1.5 evaluate to LESS THAN 1.2.0 and 1.5.0
respectively. This makes sense for sorting versions, since one expects
the versions without patch release to 'contain' the ones with, so the
ones without should be evaluated uniformily with the ones with the
patch.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln commented:
@Edgars-Cirulis I went ahead and changed the comparison functions to match FlexVer too, so we can get the test vectors working too. In the meanwhile I also found a slight bug with your implementation based on the test cases. Please see if you're fine with my changes, and feel free to make any other ones

@Ryex I've changed that behavior you've described of `1.5` == `1.5.0`, explained a bit of my reasoning in the last commit's message :-)

Ryex commented:
@flowln  Looks good, I figured that we didn't want to potentially anything by breaking from the current treatment of 1.2 vs 1.2.0 but I suppose it *is* and edge case and if it becomes a problem we can always address it elsewhere.

New test's pass for me

edgars-cirulis commented:
Looks good to me too! 👍🏼
